src/flow/persistence/mod.rs

```diff
@@ -66,13 +75,100 @@ impl Parameters {
     /// `queue_capacity` indicates the number of packets that should be buffered until
     /// flushing, and `flush_timeout` indicates the length of time to wait before flushing
     /// anyway.
-    pub fn new(mode: DurabilityMode, queue_capacity: usize, flush_timeout: time::Duration) -> Self {
+    pub fn new(
+        mode: DurabilityMode,
+        queue_capacity: usize,
+        flush_timeout: time::Duration,
+        log_prefix: Option<String>,
+    ) -> Self {
         Self {
             queue_capacity,
             flush_timeout,
             mode,
+            log_prefix: log_prefix.unwrap_or(String::from("soup")),
         }
     }
+
+    /// The path that would be used for the given domain/shard pair's logs.
+    pub fn log_path(
+        &self,
+        node: &LocalNodeIndex,
+        domain_index: domain::Index,
+        domain_shard: usize,
+    ) -> PathBuf {
+        let filename = format!(
+            "{}-log-{}_{}-{}.json",
+            self.log_prefix,
+            domain_index.index(),
+            domain_shard,
+            node.id()
+        );
+
+        PathBuf::from(&filename)
+    }
+}
+
+/// Retrieves a vector of packets from the persistent log.
+pub fn retrieve_recovery_packets(
+    nodes: &DomainNodes,
+    domain_index: domain::Index,
+    domain_shard: usize,
+    params: &Parameters,
+    checktable: Rc<checktable::CheckTableClient>,
+) -> Vec<Box<Packet>> {
+    let mut packets = vec![];
+    for (_index, node) in nodes.iter() {
+        let node = node.borrow();
+        let local_addr = node.local_addr();
+        let global_addr = node.global_addr();
+        let path = params.log_path(&local_addr, domain_index, domain_shard);
+        let file = match File::open(&path) {
+            Ok(f) => f,
+            Err(ref e) if e.kind() == ErrorKind::NotFound => continue,
+            Err(e) => panic!("Could not open log file {:?}: {}", path, e),
+        };
+
+        BufReader::new(file)
+            .lines()
+            .flat_map(|line| {
+                let line = line.unwrap();
+                let entries: Vec<Records> = serde_json::from_str(&line).unwrap();
+                entries
+            })
+            .enumerate()
+            .map(|(i, data)| {
+                let link = Link::new(*local_addr, *local_addr);
+                if node.is_transactional() {
+                    let id = checktable::TransactionId(i as u64);
+                    let transactions = vec![(id, data.clone(), None)];
+                    let request = checktable::service::TimestampRequest {
+                        transactions,
+                        base: global_addr,
+                    };
+
+                    let reply = checktable.apply_batch(request).unwrap().unwrap();
```

Comment:
It would be better to avoid getting a separate timestamp for every packet. Instead, merge packets from the same base node and assign the combination a single timestamp. Based on other experiments, ~512 packets to a batch will probably provide good efficiency.

Reply:
Sounds good. I guess it doesn't hurt to merge regular packets as well, and not just transactional ones?

Reply:
Fixed in 57cd580f1b75406e40c837e832058bbb5cf87fed